src/plugin/PluginManager.php

```diff
@@ -453,6 +456,15 @@ public function enablePlugin(Plugin $plugin) : bool{
 			if($plugin->isEnabled()){ //the plugin may have disabled itself during onEnable()
 				$this->enabledPlugins[$plugin->getDescription()->getName()] = $plugin;
 
+				foreach($plugin->getDescription()->getDepend() as $dependency){
+					$this->pluginDependents[$dependency][$plugin->getDescription()->getName()] = true;
+				}
+				foreach($plugin->getDescription()->getSoftDepend() as $dependency){
+					if(isset($this->plugins[$dependency])){
```

Comment:
could the dependency get loaded after server start, causing this to work incorrectly?

Reply:
Technically yes, but this isn't a use case we actively support anyway. If the plugin reached this point, it's because we decided to ignore the missing soft dependency and load it anyway, so the plugin shouldn't really be attempting to use it anyway.

See also: #4465

Reply:
No. That would cause plugins with unresolved dependencies to get unloaded in the wrong order.

Reply:
how would it? it is just a pseudo node to hold the place when the plugin actually gets loaded.

Reply:
Plugins won't be disabled until the number of enabled dependents reaches 0.

Having verbalized that, I can see what you mean...

Reply:
A case where your suggestion results in a problem:
- Plugin1 has a softdepend on Plugin2, but Plugin2 is not found. Softdepend is ignored, and Plugin1 is loaded anyway
- Plugin2, which circularly depends on Plugin1, is later loaded manually from outside the plugin folder
- Now, a circular dependency has been established within `pluginDependents`, and the server enter an infinite loop on shutdown

If these plugins were loaded together, they'd both fail to load due to circular dependencies. But due to the nature of softdepend, loading them separately would allow this problem to occur. That's why this check is needed.

Reply:
This issue would be removed by implementing #4465 , but that's outside the scope of this PR.

Reply:
I suppose we can neglect the problems with late-loaded plugins, since they already have their own problems with a tracking issue.